apidoc/Titanium/UI/WebView.yml

```diff
@@ -368,6 +463,69 @@ events:
       - name: url
         summary: The URL of the web document that is stopped.
 
+  - name: message
+    summary: Fired when a script message is received from a webpage.
+    description: |
+        This event get fired when you have added message handler using method 'addScriptMessageHandler'
+        and  webpage sends message to it.
+    platforms: [ipad, iphone]
+    since: "7.4.0"
+    properties:
+      - name: url
+        summary: URL of the web document being loaded.
+        type: String
+      - name: body
+        summary: The body of message sent from webview.
+        type: String
+      - name: name
+        summary: The name of the message handler to which the message is sent.
+        type: String
+      - name: isMainFrame
+        summary: |
+            A Boolean value indicating whether the frame is the web site's main frame or a subframe.
+        type: Boolean
+
+  - name: progress
+    summary: Fired when webpage download progresses.
+    description: |
+        This value ranges from 0.0 to 1.0 based on the total number of bytes expected to be received, 
+        including the main document and all of its potential subresources.
+    platforms: [ipad, iphone]
+    since: "7.4.0"
+    properties:
+      - name: value
+        summary: An estimate of what fraction of the current navigation has been loaded.
+        type: Number
+      - name: url
+        summary: URL of the web document being loaded.
+        type: String
+
+  - name: redirect
+    summary: Fired when a web view receives a server redirect.
+    platforms: [ipad, iphone]
+    since: "7.4.0"
+    properties:
+      - name: title
+        summary: Page title of webpage.
+        type: String
+      - name: url
+        summary: URL of the web document being loaded.
+        type: String
```

Comment:
Does "redirect" need to be its own event? Can we consolidate it with the "handleurl" event instead and provide a property indicating what triggered it?

Reply:
"handleurl" is only for handling url-schemes, e.g. `fb://`. How does Android signalize redirects? Via beforeload?

Reply:
On Android 6.x and older OS versions doesn't distinguish. All URL requests are passed to the Java [WebViewClient.shouldOverrideUrlLoading()](https://developer.android.com/reference/android/webkit/WebViewClient.html#shouldOverrideUrlLoading(android.webkit.WebView,%20java.lang.String)) method. This includes redirects and custom URL schemes. The only way to know if it's a redirect is for us track the URL usage ourselves, which we can do. Unfortunately, this method never gets called when loading URLs via the web page's JavaScript code, which is a bummer. We can only detect JavaScript loaded URLs after the HTTP request has been sent via the [WebViewClient.onPageStarted()](https://developer.android.com/reference/android/webkit/WebViewClient.html#onPageStarted(android.webkit.WebView,%20java.lang.String,%20android.graphics.Bitmap)) method. At least that's how I remember it back in the Android 2.x/3.x days.

Android 7.0 has a new [WebResourceRequest](https://developer.android.com/reference/android/webkit/WebResourceRequest) to determine if it's a redirect, but if we track it ourselves for Android 6.x and below, then there is no point in using it.